plonky2/src/recursion/cyclic_recursion.rs

```diff
@@ -0,0 +1,490 @@
+        }
+
+        let dummy_verifier_data = VerifierCircuitTarget {
+            constants_sigmas_cap: self.add_virtual_cap(self.config.fri_config.cap_height),
+            circuit_digest: self.add_virtual_hash(),
+        };
+
+        // Flag set to true for the base case of the cycle where we verify a dummy proof to bootstrap the cycle. Set to false otherwise.
+        // Unsafe is ok since `base_case` is a public input and its booleaness should be checked in the verifier.
+        let base_case = self.add_virtual_bool_target_unsafe();
+        self.register_public_input(base_case.target);
+
+        common_data.num_public_inputs = self.num_public_inputs();
+        // The `conditionally_verify_proof` gadget below takes 2^12 gates, so `degree_bits` cannot be smaller than 13.
+        common_data.fri_params.degree_bits = common_data.fri_params.degree_bits.max(13);
+
+        let proof = self.add_virtual_proof_with_pis(&common_data);
+        let dummy_proof = self.add_virtual_proof_with_pis(&common_data);
+
+        let pis = CyclicPublicInputsTarget::from_slice(&proof.public_inputs, &common_data)?;
+        // Check that the previous base case flag was boolean.
+        self.assert_bool(pis.base_case);
+        // Check that we cannot go from a non-base case to a base case by checking `previous_base_case - base_case \in {0,1}`.
+        let decrease = BoolTarget::new_unsafe(self.sub(pis.base_case.target, base_case.target));
+        self.assert_bool(decrease);
+        // Connect previous verifier data to current one. This guarantees that every proof in the cycle uses the same verifier data.
+        self.connect_hashes(pis.circuit_digest, verifier_data.circuit_digest);
+        for (h0, h1) in pis
+            .constants_sigmas_cap
+            .0
+            .iter()
+            .zip_eq(&verifier_data.constants_sigmas_cap.0)
+        {
+            self.connect_hashes(*h0, *h1);
+        }
+
+        self.connect(previous_base_case, pis.base_case.target);
+        for (x, y) in previous_virtual_public_inputs
+            .iter()
+            .zip(&proof.public_inputs)
+        {
+            self.connect(*x, *y);
+        }
+
+        // Verify the dummy proof if `base_case` is set to true, otherwise verify the "real" proof.
+        self.conditionally_verify_proof(
+            base_case,
+            &dummy_proof,
+            &dummy_verifier_data,
+            &proof,
+            &verifier_data,
+            &common_data,
+        );
+
+        // Make sure we have enough gates to match `common_data`.
+        while self.num_gates() < (common_data.degree() / 2) {
+            self.add_gate(NoopGate, vec![]);
+        }
+        // Make sure we have every gate to match `common_data`.
+        for g in &common_data.gates {
+            self.add_gate_to_gate_set(g.clone());
+        }
+
+        let data = self.build::<C>();
+        ensure!(data.common == common_data, "Common data does not match.");
```

Comment:
nit: let's include the two objects and maybe clarify a bit that the `common_data` passed in didn't match the final circuit. I think this will be a common error since users need to guess at `common_data`